speculation-rules/prerender/resources/service-worker.js

```diff
@@ -0,0 +1,11 @@
+self.addEventListener("fetch", e => {
+    if (e.request.url.endsWith("ping"))
+        e.respondWith(new Response('pong'));
+    else if (e.request.url.endsWith("client")) {
+        e.respondWith(new Promise(async resolve => {
```

Comment:
nit: if you're going to use `async/await` anyway, surely that totally avoids the need for the `Promise` ctor at all and rolls errors forward cleanly:

```javascript
e.respondWith((async () => {
    const client = await clients.get(e.clientId);
    const clientInfo = ...;
    return new Response(...);
})());
```